neqo-transport/src/cc/tests/cubic.rs

```diff
@@ -0,0 +1,301 @@
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![allow(clippy::cast_possible_truncation)]
+#![allow(clippy::cast_sign_loss)]
+
+use crate::cc::{
+    classic_cc::{ClassicCongestionControl, CWND_INITIAL},
+    cubic::{
+        Cubic, CUBIC_ALPHA, CUBIC_BETA_USIZE_DIVISOR, CUBIC_BETA_USIZE_QUOTIENT, CUBIC_C,
+        CUBIC_FAST_CONVERGENCE,
+    },
+    CongestionControl, MAX_DATAGRAM_SIZE, MAX_DATAGRAM_SIZE_F64,
+};
+use crate::packet::PacketType;
+use crate::tracking::SentPacket;
+use std::convert::TryFrom;
+use std::time::{Duration, Instant};
+use test_fixture::now;
+
+const RTT: Duration = Duration::from_millis(100);
+const CWND_INITIAL_F64: f64 = 10.0 * MAX_DATAGRAM_SIZE_F64;
+const CWND_INITIAL_10_F64: f64 = 10.0 * CWND_INITIAL_F64;
+const CWND_INITIAL_10: usize = 10 * CWND_INITIAL;
+const CWND_AFTER_LOSS: usize = CWND_INITIAL * CUBIC_BETA_USIZE_QUOTIENT / CUBIC_BETA_USIZE_DIVISOR;
+const CWND_AFTER_LOSS_SLOW_START: usize =
+    (CWND_INITIAL + MAX_DATAGRAM_SIZE) * CUBIC_BETA_USIZE_QUOTIENT / CUBIC_BETA_USIZE_DIVISOR;
+
+fn fill_cwnd(cc: &mut ClassicCongestionControl<Cubic>, mut next_pn: u64, now: Instant) -> u64 {
+    while cc.bytes_in_flight() < cc.cwnd() {
+        let sent = SentPacket::new(
+            PacketType::Short,
+            next_pn,           // pn
+            now,               // time sent
+            true,              // ack eliciting
+            Vec::new(),        // tokens
+            MAX_DATAGRAM_SIZE, // size
+        );
+        cc.on_packet_sent(&sent);
+        next_pn += 1;
+    }
+    next_pn
+}
+
+fn ack_packet(cc: &mut ClassicCongestionControl<Cubic>, pn: u64, now: Instant) {
+    let acked = SentPacket::new(
+        PacketType::Short,
+        pn,                // pn
+        now,               // time sent
+        true,              // ack eliciting
+        Vec::new(),        // tokens
+        MAX_DATAGRAM_SIZE, // size
+    );
+    cc.on_packets_acked(&[acked], RTT, now);
+}
+
+fn packet_lost(cc: &mut ClassicCongestionControl<Cubic>, pn: u64) {
+    const PTO: Duration = Duration::from_millis(120);
+    let p_lost = SentPacket::new(
+        PacketType::Short,
+        pn,                // pn
+        now(),             // time sent
+        true,              // ack eliciting
+        Vec::new(),        // tokens
+        MAX_DATAGRAM_SIZE, // size
+    );
+    cc.on_packets_lost(None, None, PTO, &[p_lost]);
+}
+
+#[test]
+fn tcp_phase() {
```

Comment:
This test does too much.  It tests that the TCP phase works, then it tests that the increases is faster than TCP when it exits, then it tests that it increases at precisely the cubic rate.

This is at least three tests.

Reply:
it is not always easy to construct the starting point of a test (we will need to expose more functions to modify cc internal values). 
I wrote the test to show that: tcp phase is exited exactly at the point expected and that the increase after that is not using tcp equation, but cubic equation. Therefore you need all 3 pieces in one test.

Reply:
I think that I can live with this.  It's not ideal, but it's also not that long relative to some of the other tests.